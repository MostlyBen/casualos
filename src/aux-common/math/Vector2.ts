/**
 * Defines a class that represents a 2D point in space.
 */
export class Vector2 {
    /**
     * The X value of this vector.
     */
    x: number;

    /**
     * The Y value of this vector.
     */
    y: number;

    /**
     * Constructs a new 2D vector with the given X and Y values.
     * @param x The X value of the vector.
     * @param y The Y value of the vector.
     */
    constructor(x: number = 0, y: number = 0) {
        this.x = x;
        this.y = y;
    }

    /**
     * Creates a 2D vector with the given X and Y values that is normalized immediately upon creation.
     * @param x The X value of the vector.
     * @param y The Y value of the vector.
     */
    static createNormalized(x: number, y: number) {
        const length = Math.sqrt(x * x + y * y);
        return new Vector2(x / length, y / length);
    }

    /**
     * Calculates the angle between the two given vectors and returns the result in radians.
     * @param first The first vector.
     * @param second The second vector.
     */
    static angleBetween(first: Vector2, second: Vector2): number {
        const dot = first.dot(second);
        const l1 = first.length();
        const l2 = second.length();
        return Math.acos(dot / (l1 * l2));
    }

    /**
     * Calculates the distance between the two given vectors and returns the result.
     * @param first The first vector.
     * @param second The second vector.
     */
    static distanceBetween(first: Vector2, second: Vector2): number {
        const direction = second.subtract(first);
        return direction.length();
    }

    /**
     * Adds this vector with the other vector and returns the result.
     * @param other The other vector to add with this vector.
     */
    add(other: Vector2): Vector2 {
        return new Vector2(this.x + other.x, this.y + other.y);
    }

    /**
     * Subtracts the other vector from this vector and returns the result.
     * @param other The other vector that should be subtracted from this vector.
     */
    subtract(other: Vector2): Vector2 {
        return new Vector2(this.x - other.x, this.y - other.y);
    }

    /**
     * Multiplies each component of this vector by the given value and returns the result.
     * @param scale The scale that should be applied to this vector.
     */
    multiplyScalar(scale: number): Vector2 {
        return new Vector2(this.x * scale, this.y * scale);
    }

    /**
     * Multiplies this vector by the given other vector and returns the result.
     * @param other The other vector to multiply with this vector.
     */
    multiply(other: Vector2): Vector2 {
        return new Vector2(this.x * other.x, this.y * other.y);
    }

    /**
     * Calculates the dot product of this vector compared to the given other vector.
     * Returns a number that is positive if the vectors point in the same direction,
     * negative if they point in opposite directions, and zero if they are perpendicular.
     * For normalized vectors, this value is clamped to 1 and -1.
     * @param other The other vector to calculate the dot product with.
     */
    dot(other: Vector2): number {
        return this.x * other.x + this.y * other.y;
    }

    /**
     * Calculates the length of this vector and returns the result.
     */
    length(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    /**
     * Calculates the square length of this vector and returns the result.
     * This is equivalent to length^2, but it is faster to calculate than length because it doesn't require
     * calculating a square root.
     */
    squareLength(): number {
        return this.x * this.x + this.y * this.y;
    }

    /**
     * Calculates the normalized version of this vector and returns it.
     * A normalized vector is a vector whose length equals 1.
     *
     * Normalizing a vector preserves its directionality while making the length (i.e. scale) of it 1.
     */
    normalize(): Vector2 {
        const length = this.length();
        return new Vector2(this.x / length, this.y / length);
    }

    toString(): string {
        return `Vector2(${this.x}, ${this.y})`;
    }

    /**
     * Determines if this vector equals the other vector.
     * @param other The other vector.
     */
    equals(other: Vector2): boolean {
        return this.x === other.x && this.y === other.y;
    }
}
