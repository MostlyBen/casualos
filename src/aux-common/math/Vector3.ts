import { clamp } from '../utils';
import { Vector2 } from './Vector2';

/**
 * Defines a class that represents a 3D point in space.
 */
export class Vector3 {
    /**
     * The X value of this vector.
     */
    x: number;

    /**
     * The Y value of this vector.
     */
    y: number;

    /**
     * The Z value of this vector.
     */
    z: number;

    /**
     * Gets a new Vector2 that contains this vector's X and Y components.
     */
    get xy() {
        return new Vector2(this.x, this.y);
    }

    /**
     * Gets a new Vector2 that contains this vector's X and Z components.
     */
    get xz() {
        return new Vector2(this.x, this.z);
    }

    /**
     * Gets a new Vector2 that contains this vector's Y and Z components.
     */
    get yz() {
        return new Vector2(this.y, this.z);
    }

    /**
     * Constructs a new 3D vector with the given X and Y values.
     * @param x The X value of the vector.
     * @param y The Y value of the vector.
     * @param z The Z value of the vector.
     */
    constructor(x: number = 0, y: number = 0, z: number = 0) {
        this.x = x;
        this.y = y;
        this.z = z;
        Object.freeze(this);
    }

    /**
     * Creates a 3D vector with the given X and Y values that is normalized immediately upon creation.
     * @param x The X value of the vector.
     * @param y The Y value of the vector.
     * @param z The Z value of the vector.
     */
    static createNormalized(x: number, y: number, z: number) {
        const length = Math.sqrt(x * x + y * y + z * z);
        return new Vector3(x / length, y / length, z / length);
    }

    /**
     * Calculates the angle between the two given vectors and returns the result in radians.
     * @param first The first vector that should be used for comparision.
     * @param second The second vector that should be used for comparision.
     */
    static angleBetween(first: Vector3, second: Vector3): number {
        const dot = first.dot(second);
        const l1 = first.length();
        const l2 = second.length();
        const cos = dot / (l1 * l2);
        if (cos <= 1 && cos >= -1) {
            return Math.acos(cos);
        } else {
            // Sometimes the dot product ends up outside the 1 <-> -1 range and we need to clamp it.
            return Math.acos(clamp(cos, -1, 1));
        }
    }

    /**
     * Calculates the distance between the two given vectors and returns the result.
     * @param first The first vector that should be used for comparision.
     * @param second The second vector that should be used for comparision.
     */
    static distanceBetween(first: Vector3, second: Vector3): number {
        const direction = second.subtract(first);
        return direction.length();
    }

    /**
     * Constructs a new vector that is the linear interpolation between the given start and end positions.
     * The degree that the result is interpolated is determined by the given amount parameter.
     * @param start The start position.
     * @param finish The end position.
     * @param amount The amount that the resulting position should be interpolated between the start and end positions.  Values near 0 indicate rotations close to the first and values near 1 indicate rotations close to the second.
     */
    static interpolatePosition(
        start: Vector3,
        finish: Vector3,
        amount: number
    ) {
        const dir = finish.subtract(start);
        const lerp = dir.multiplyScalar(amount);
        return start.add(lerp);
    }

    /**
     * Constructs a new vector that is the directional linear interpolation between the given start and end positions.
     * The degree that the result is interpolated is determined by the given amount parameter.
     *
     * This function works similarly to interpolatePosition(), except the result is always a normalized vector.
     *
     * @param start The start position.
     * @param finish The end position.
     * @param amount The amount that the resulting position should be interpolated between the start and end positions.  Values near 0 indicate rotations close to the first and values near 1 indicate rotations close to the second.
     */
    static interpolateDirection(
        start: Vector3,
        finish: Vector3,
        amount: number
    ) {
        return Vector3.interpolatePosition(start, finish, amount).normalize();
    }

    /**
     * Adds this vector with the other vector and returns the result.
     * @param other The other vector to add with this vector.
     */
    add(other: Vector3): Vector3 {
        return new Vector3(
            this.x + other.x,
            this.y + other.y,
            this.z + other.z
        );
    }

    /**
     * Subtracts the other vector from this vector and returns the result.
     * @param other The other vector that should be subtracted from this vector.
     */
    subtract(other: Vector3): Vector3 {
        return new Vector3(
            this.x - other.x,
            this.y - other.y,
            this.z - other.z
        );
    }

    /**
     * Multiplies each component of this vector by the given value and returns the result.
     * @param scale The scale that should be applied to this vector.
     */
    multiplyScalar(scale: number): Vector3 {
        return new Vector3(this.x * scale, this.y * scale, this.z * scale);
    }

    /**
     * Multiplies this vector by the given other vector and returns the result.
     * @param other The other vector to multiply with this vector.
     */
    multiply(other: Vector3): Vector3 {
        return new Vector3(
            this.x * other.x,
            this.y * other.y,
            this.z * other.z
        );
    }

    /**
     * Calculates the dot product of this vector compared to the given other vector.
     * Returns a number that is positive if the vectors point in the same direction,
     * negative if they point in opposite directions, and zero if they are perpendicular.
     * For normalized vectors, this value is clamped to 1 and -1.
     * @param other The other vector to calculate the dot product with.
     */
    dot(other: Vector3): number {
        return this.x * other.x + this.y * other.y + this.z * other.z;
    }

    /**
     * Calculates the cross product of this vector with the given other vector.
     * Returns a new vector that is perpendicular to both vectors.
     * Note that the order of the vectors greatly matters. For example, (1, 0, 0).cross(0, 1, 0) === (0, 0, 1) but (0, 1, 0).cross(1, 0, 0) === (0, 0, -1).
     * @param other The other vector to calculate the cross product with.
     */
    cross(other: Vector3): Vector3 {
        return new Vector3(
            this.y * other.z - other.y * this.z,
            this.z * other.x - other.z * this.x,
            this.x * other.y - other.x * this.y
        );
    }

    /**
     * Calculates the length of this vector and returns the result.
     */
    length(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
    }

    /**
     * Calculates the square length of this vector and returns the result.
     * This is equivalent to length^2, but it is faster to calculate than length because it doesn't require
     * calculating a square root.
     */
    squareLength(): number {
        return this.x * this.x + this.y * this.y + this.z * this.z;
    }

    /**
     * Calculates the normalized version of this vector and returns it.
     * A normalized vector is a vector whose length equals 1.
     */
    normalize(): Vector3 {
        const length = this.length();
        if (length === 1) {
            return this;
        }
        return new Vector3(this.x / length, this.y / length, this.z / length);
    }

    /**
     * Negates each component of this vector and returns a new vector that contains the result.
     */
    negate(): Vector3 {
        return new Vector3(-this.x, -this.y, -this.z);
    }

    toString(): string {
        return `Vector3(${this.x}, ${this.y}, ${this.z})`;
    }

    /**
     * Determines if this vector equals the other vector.
     * @param other The other value to compare to.
     */
    equals(other: Vector3): boolean {
        return this.x === other.x && this.y === other.y && this.z === other.z;
    }
}

/**
 * A 3D vector that contains (0, 0, 0).
 */
export const ZERO = new Vector3();

/**
 * A 3D vector that contains (1, 1, 1).
 */
export const ONE = new Vector3(1, 1, 1);

/**
 * A 3D vector that contains (0, 1, 0).
 */
export const FORWARD = new Vector3(0, 1, 0);

/**
 * A 3D vector that contains (0, -1, 0).
 */
export const BACK = new Vector3(0, -1, 0);

/**
 * A 3D vector that contains (1, 0, 0).
 */
export const RIGHT = new Vector3(1, 0, 0);

/**
 * A 3D vector that contains (-1, 0, 0).
 */
export const LEFT = new Vector3(-1, 0, 0);

/**
 * A 3D vector that contains (0, 0, 1).
 */
export const UP = new Vector3(0, 0, 1);

/**
 * A 3D vector that contains (0, 0, -1).
 */
export const DOWN = new Vector3(0, 0, -1);
