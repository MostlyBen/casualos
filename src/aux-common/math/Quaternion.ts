import { Vector2 } from './Vector2';
import { Vector3 } from './Vector3';

/**
 * Defines a class that represents a Quaternion. That is, a 3D rotation.
 *
 * Quaternions are a mathematical representation of 3D transformations and are commonly used to calculate and apply rotations to 3D points.
 * They work by defining a quaterion such that q = w + x*i + y*j + z*k, where w, x, y, and z are real numbers and i, j, and k are imaginary numbers.
 * The basics of this is that x, y, and z define a vector that represents the rotation axis, and w defines an angle around which the rotation occurs.
 * However, because i, j, and k are included we can keep x, y, and z from incorrectly interacting with each other and so avoid common pitfalls like Gimbal lock.
 *
 * One little known feature of quaternions is that they can also represent reflections and also scale.
 * This is because there are two different ways to apply a quaternion to a 3D point:
 *
 * - quaterion * point * inverse(quaterion)
 *
 * This formula rotates and scales the point quaternion. The rotation occurs around the axis specified by the quaternion X, Y, and Z values.
 * Additionally, the point will be scaled by the length of the quaternion. (i.e. sqrt( x^2 + y^2 + z^2 + w^2 ))
 * This is why quaternions that are used to represent only rotations must be normalized.
 *
 * - quaternion * point * quaternion
 *
 * This formula reflects scales the point by the quaternion. The reflection occurs across the axis specified by the quaternion X, Y, and Z values.
 * Additionally, the point will be scaled by the length of the quaternion. (i.e. sqrt( x^2 + y^2 + z^2 + w^2 ))
 */
export class Quaternion {
    /**
     * The X value of the quaternion.
     */
    x: number;

    /**
     * The Y value of the quaternion.
     */
    y: number;

    /**
     * The Z value of the quaternion.
     */
    z: number;

    /**
     * The W value of the quaternion.
     */
    w: number;

    /**
     * Creates a new Quaternion with the given values.
     * @param x The X value.
     * @param y The Y value.
     * @param z The Z value.
     * @param w The W value.
     */
    constructor(x: number = 0, y: number = 0, z: number = 0, w: number = 1) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.w = w;
    }

    /**
     * Constructs a new Quaternion from the given axis and angle.
     * @param axisAndAngle The object that contains the axis and angle values.
     */
    static rotationFromAxisAndAngle(axisAndAngle: AxisAndAngle): Quaternion {
        const normalizedAxis = axisAndAngle.axis.normalize();
        const sinAngle = Math.sin(axisAndAngle.angle / 2);
        const cosAngle = Math.cos(axisAndAngle.angle / 2);
        return new Quaternion(
            normalizedAxis.x * sinAngle,
            normalizedAxis.y * sinAngle,
            normalizedAxis.z * sinAngle,
            cosAngle
        );
    }

    /**
     * Determines the angle between the two given quaternions and returns the result in radians.
     * @param first The first quaternion. Must be a quaterion that represents a rotation
     * @param second The second quaternion.
     */
    static angleBetweenRotations(
        first: Quaternion,
        second: Quaternion
    ): number {
        const delta = first.invert().multiply(second);
        return 2 * Math.acos(delta.w);
    }

    /**
     * Rotates the given Vector3 by this quaternion and returns a new vector containing the result.
     * @param vector The 3D vector that should be rotated.
     */
    rotateVector3(vector: Vector3): Vector3 {
        // Multiplied out version of (q * vector * q^-1)
        return new Vector3(
            this.w * this.w * vector.x +
                2 * this.y * this.w * vector.z -
                2 * this.z * this.w * vector.y +
                this.x * this.x * vector.x +
                2 * this.y * this.x * vector.y +
                2 * this.z * this.x * vector.z -
                this.z * this.z * vector.x -
                this.y * this.y * vector.x,
            2 * this.x * this.y * vector.x +
                this.y * this.y * vector.y +
                2 * this.z * this.y * vector.z +
                2 * this.w * this.z * vector.x -
                this.z * this.z * vector.y +
                this.w * this.w * vector.y -
                2 * this.x * this.w * vector.z -
                this.x * this.x * vector.y,
            2 * this.x * this.z * vector.x +
                2 * this.y * this.z * vector.y +
                this.z * this.z * vector.z -
                2 * this.w * this.y * vector.x -
                this.y * this.y * vector.z +
                2 * this.w * this.x * vector.y -
                this.x * this.x * vector.z +
                this.w * this.w * vector.z
        );
    }

    /**
     * Rotates the given Vector2 by this quaternion and returns a new vector containing the result.
     * Note that rotations around any other axis than (0, 0, 1) or (0, 0, -1) can produce results that contain a Z component that is dropped because the result is a Vector2.
     * @param vector The 2D vector that should be rotated.
     */
    rotateVector2(vector: Vector2): Vector2 {
        // Multiplied out version of (q * vector * q^-1)
        return new Vector2(
            this.w * this.w * vector.x -
                2 * this.z * this.w * vector.y +
                this.x * this.x * vector.x +
                2 * this.y * this.x * vector.y -
                this.z * this.z * vector.x -
                this.y * this.y * vector.x,
            2 * this.x * this.y * vector.x +
                this.y * this.y * vector.y +
                2 * this.w * this.z * vector.x -
                this.z * this.z * vector.y +
                this.w * this.w * vector.y -
                this.x * this.x * vector.y
        );
    }

    /**
     * Multiplies this quaternion by the other quaternion and returns the result.
     * In quaternion math, multiplication can be used to combine quaternions together,
     * however unlike regular multiplication quaternion multiplication is order dependent.
     *
     * Which frame of reference you want to use depends on which order you use.
     * For example, q2.multiply(q1) starts with the identity, applies q1 to it, and then applies q2 to that.
     * Whereas, q1.multiply(q2) starts with the identity, applies q2 to it, and then applies q1 to that.
     *
     * @param other The other quaternion.
     */
    multiply(other: Quaternion): Quaternion {
        // Taken from https://www.euclideanspace.com/maths/algebra/realNormedAlgebra/quaternions/arithmetic/index.htm
        const x =
            this.x * other.w +
            this.w * other.x +
            this.y * other.z -
            this.z * other.y;
        const y =
            this.w * other.y -
            this.x * other.z +
            this.y * other.w +
            this.z * other.x;
        const z =
            this.w * other.z +
            this.x * other.y -
            this.y * other.x +
            this.z * other.w;
        const w =
            this.w * other.w -
            this.x * other.x -
            this.y * other.y -
            this.z * other.z;
        return new Quaternion(x, y, z, w);
    }

    /**
     * Calculates the conjugate of this quaternion and returns the result.
     * The conjugate (or inverse) of a quaternion is similar to negating a number.
     * When you multiply a quaternion by its conjugate, the result is the identity quaternion.
     */
    invert(): Quaternion {
        return new Quaternion(-this.x, -this.y, -this.z, this.w);
    }

    /**
     * Gets the length of this vector. That is, the pathagorean theorem applied to X, Y, Z, and W.
     */
    length(): number {
        return Math.sqrt(
            this.x * this.x +
                this.y * this.y +
                this.z * this.z +
                this.w * this.w
        );
    }

    /**
     * Calculates the square length of this quaternion and returns the result.
     * This is equivalent to length^2, but it is faster to calculate than length because it doesn't require
     * calculating a square root.
     */
    squareLength(): number {
        return (
            this.x * this.x +
            this.y * this.y +
            this.z * this.z +
            this.w * this.w
        );
    }

    /**
     * Calculates the normalized version of this quaternion and returns it.
     * A normalized quaternion is a quaternion whose length equals 1.
     *
     * Normalizing a quaternion preserves its rotation/reflection while making the length (i.e. scale) of it 1.
     */
    normalize(): Quaternion {
        const length = this.length();
        return new Quaternion(
            this.x / length,
            this.y / length,
            this.z / length,
            this.w / length
        );
    }

    toString(): string {
        return `Quaternion(${this.x}, ${this.y}, ${this.z}, ${this.w})`;
    }

    /**
     * Determines if this quaternion equals the other quaternion.
     * @param other The other quaternion to apply.
     */
    equals(other: Quaternion): boolean {
        if (!other) {
            return false;
        }

        return (
            this.x === other.x &&
            this.y === other.y &&
            this.z === other.z &&
            this.w === other.w
        );
    }
}

/**
 * Defines an interface that represents an Axis and Angle pair.
 */
export interface AxisAndAngle {
    /**
     * The axis about which the angle should rotate around.
     */
    axis: Vector3;

    /**
     * The number of radians that should be rotated around the axis.
     */
    angle: number;
}
