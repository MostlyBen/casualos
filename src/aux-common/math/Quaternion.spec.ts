import { MathUtils } from 'three';
import { Quaternion } from './Quaternion';
import { Vector2 } from './Vector2';
import { Vector3 } from './Vector3';

describe('Quaternion', () => {
    it('should default to (0, 0, 0, 1)', () => {
        const q = new Quaternion();
        expect(q.x).toBe(0);
        expect(q.y).toBe(0);
        expect(q.z).toBe(0);
        expect(q.w).toBe(1);
    });

    it('should contain x, y, z, and w values', () => {
        const q = new Quaternion(1, 2, 3, 4);
        expect(q.x).toBe(1);
        expect(q.y).toBe(2);
        expect(q.z).toBe(3);
        expect(q.w).toBe(4);
    });

    describe('toString()', () => {
        it('should return a nicely formatted string', () => {
            const q = new Quaternion(1, 2, 3, 4);

            expect(q.toString()).toBe('Quaternion(1, 2, 3, 4)');
        });
    });

    describe('equals()', () => {
        it('should return false if given null', () => {
            const q1 = new Quaternion();

            expect(q1.equals(null)).toBe(false);
        });

        it('should return true if the quaternions are equal', () => {
            const q1 = new Quaternion(1, 2, 3, 4);
            const q2 = new Quaternion(1, 2, 3, 4);

            expect(q1.equals(q2)).toBe(true);
        });

        it('should return false if the quaternions differ in X', () => {
            const q1 = new Quaternion(1, 2, 3, 4);
            const q2 = new Quaternion(9, 2, 3, 4);

            expect(q1.equals(q2)).toBe(false);
        });

        it('should return false if the quaternions differ in Y', () => {
            const q1 = new Quaternion(1, 2, 3, 4);
            const q2 = new Quaternion(1, 9, 3, 4);

            expect(q1.equals(q2)).toBe(false);
        });

        it('should return false if the quaternions differ in Z', () => {
            const q1 = new Quaternion(1, 2, 3, 4);
            const q2 = new Quaternion(1, 2, 9, 4);

            expect(q1.equals(q2)).toBe(false);
        });

        it('should return false if the quaternions differ in W', () => {
            const q1 = new Quaternion(1, 2, 3, 4);
            const q2 = new Quaternion(1, 2, 3, 9);

            expect(q1.equals(q2)).toBe(false);
        });
    });

    describe('rotationFromAxisAndAngle()', () => {
        it('should construct a rotation quaternion from an axis and angle', () => {
            const q1 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 1, 0),
                angle: Math.PI / 2,
            }); // 90 degree rotation about Y axis

            expect(q1).toMatchSnapshot();
        });
    });

    describe('angleBetweenRotations()', () => {
        it('should return the angle between the two quaternions in radians', () => {
            const q1 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 1, 0),
                angle: Math.PI / 2,
            }); // 90 degree rotation about Y axis

            const q2 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 1, 0),
                angle: Math.PI / 4,
            }); // 45 degree rotation about Y axis

            expect(Quaternion.angleBetweenRotations(q1, q2)).toBeCloseTo(
                Math.PI / 4,
                5
            ); // 45 degrees
        });

        it('should return 0 if given the same quaternion', () => {
            const q1 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 1, 0),
                angle: Math.PI / 2,
            }); // 90 degree rotation about Y axis

            expect(Quaternion.angleBetweenRotations(q1, q1)).toBeCloseTo(0, 5); // 0 degrees
        });
    });

    describe('rotateVector3()', () => {
        it('should return a new point rotated by the given quaternion', () => {
            const q1 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 1, 0),
                angle: Math.PI / 2,
            }); // 90 degree rotation about Y axis

            const q2 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 0, 1),
                angle: Math.PI / 2,
            }); // 90 degree rotation about Z axis

            const p1 = q1.rotateVector3(new Vector3(1, 0, 0));
            expect(p1.x).toBeCloseTo(0);
            expect(p1.y).toBeCloseTo(0);
            expect(p1.z).toBeCloseTo(-1);

            const p2 = q2.rotateVector3(new Vector3(1, 0, 0));
            expect(p2.x).toBeCloseTo(0);
            expect(p2.y).toBeCloseTo(1);
            expect(p2.z).toBeCloseTo(0);
        });
    });

    describe('rotateVector2()', () => {
        it('should return a new point rotated by the given quaternion', () => {
            const q1 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 1, 0),
                angle: Math.PI / 2,
            }); // 90 degree rotation about Y axis

            const q2 = Quaternion.rotationFromAxisAndAngle({
                axis: new Vector3(0, 0, 1),
                angle: Math.PI / 2,
            }); // 90 degree rotation about Z axis

            const p1 = q1.rotateVector2(new Vector2(1, 0));
            expect(p1.x).toBeCloseTo(0);
            expect(p1.y).toBeCloseTo(0);

            const p2 = q2.rotateVector2(new Vector2(1, 0));
            expect(p2.x).toBeCloseTo(0);
            expect(p2.y).toBeCloseTo(1);
        });
    });

    describe('multiply()', () => {
        it('should multiply the quaternions together', () => {
            const q1 = new Quaternion(1, 2, 3, 4);
            const q2 = new Quaternion(5, 6, 7, 8);

            const q3 = q1.multiply(q2);

            expect(q3.x).toBe(24);
            expect(q3.y).toBe(48);
            expect(q3.z).toBe(48);
            expect(q3.w).toBe(-6);
        });
    });

    describe('invert()', () => {
        it('should return the conjugate of the quaternion', () => {
            const q1 = new Quaternion(1, 2, 3, 4);

            const q2 = q1.invert();
            expect(q2.x).toBe(-1);
            expect(q2.y).toBe(-2);
            expect(q2.z).toBe(-3);
            expect(q2.w).toBe(4);
        });
    });

    describe('length()', () => {
        it('should return length of the quaternion', () => {
            const q = new Quaternion(1, 3, 7, 11);
            expect(q.length()).toBe(Math.sqrt(1 * 1 + 3 * 3 + 7 * 7 + 11 * 11));
        });
    });

    describe('squareLength()', () => {
        it('should return length of the quaternion without the square root', () => {
            const q = new Quaternion(1, 3, 7, 11);
            expect(q.squareLength()).toBe(1 * 1 + 3 * 3 + 7 * 7 + 11 * 11);
        });
    });

    describe('normalize()', () => {
        it('should return a normalized copy of the quaternion', () => {
            const q1 = new Quaternion(1, 2, 3, 4);

            const q2 = q1.normalize();
            expect(q2.length()).toBeCloseTo(1, 5);
        });
    });
});
