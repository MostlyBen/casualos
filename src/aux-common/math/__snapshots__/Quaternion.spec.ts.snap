// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quaternion rotationFromAxisAndAngle() should construct a rotation quaternion from an axis and angle 1`] = `
Quaternion {
  "w": 0.7071067811865476,
  "x": 0,
  "y": 0.7071067811865475,
  "z": 0,
}
`;
